crypto/s2n_hash.h

```diff
@@ -63,13 +66,34 @@ struct s2n_hash_state {
             MD5_CTX md5;
             SHA_CTX sha1;
         } md5_sha1;
-    } hash_ctx;
+    } low_level;
+    struct s2n_hash_evp_digest {
+        struct s2n_evp_digest primary;
+        struct s2n_evp_digest md5_secondary;
+    } evp;
 };
 
-extern int s2n_hash_digest_size(s2n_hash_algorithm alg, uint8_t *out);
+struct s2n_hash_state {
+    s2n_hash_algorithm alg;
+    const struct s2n_hash *hash_impl;
+    union s2n_hash_digest digest;
+};
+
+struct s2n_hash {
```

Comment:
I have re-worked the structure of s2n_hash_state, removed the internal use of s2n_hash (function pointers) and abstracted that to a global variable, and re-worked naming. Let me know if that helps.